minorminer/utils/zephyr/survey.py

```diff
@@ -0,0 +1,292 @@
+                f"Expected 'int' or 'coordinate' for self.coord, got {self._input_coord_type}"
+            )
+
+    @property
+    def shape(self) -> ZShape:
+        """Returns the ZShape of G"""
+        return self._shape
+
+    @property
+    def nodes(self) -> set[ZNode]:
+        """Returns the ZNodes of the sampler/graph"""
+        return self._nodes
+
+    @cached_property
+    def missing_nodes(self) -> set[ZNode]:
+        """Returns the ZNodes of the sampler/graph which are missing compared to perfect yield
+        Zephyr graph on the same shape.
+        """
+        parent_nodes = [
+            ZNode(coord=ZephyrCoord(*v), shape=self._shape)
+            for v in dnx.zephyr_graph(m=self._shape.m, t=self._shape.t, coordinates=True).nodes()
+        ]
+        return {v for v in parent_nodes if not v in self._nodes}
+
+    @property
+    def edges(self) -> set[ZEdge]:
+        """Returns the ZEdges of the sampler/graph"""
+        return self._edges
+
+    @cached_property
+    def missing_edges(self) -> set[ZEdge]:
+        """Returns the ZEdges of the sampler/graph which are missing compared to
+        perfect yield Zephyr graph on the same shape."""
+        parent_edges = [
+            ZEdge(ZNode(coord=u, shape=self.shape), ZNode(coord=v, shape=self.shape))
+            for (u, v) in dnx.zephyr_graph(
+                m=self._shape.m, t=self._shape.t, coordinates=True
+            ).edges()
+        ]
+        return {e for e in parent_edges if not e in self._edges}
+
+    @property
+    def extra_missing_edges(self) -> set[ZEdge]:
+        """Returns the ZEdges of the sampler/graph which are missing compared to
+        perfect yield Zephyr graph on the same shape and are not incident with
+        a missing node."""
+        return {e for e in self.missing_edges if e[0] in self._nodes and e[1] in self._nodes}
+
+    @property
+    def num_nodes(self) -> int:
+        """Returns the number of nodes"""
+        return len(self._nodes)
+
+    @property
+    def num_missing_nodes(self) -> int:
+        """Returns the number of missing nodes"""
+        return len(self.missing_nodes)
+
+    @property
+    def num_edges(self) -> int:
+        """Returns the number of edges"""
+        return len(self._edges)
+
+    @property
+    def num_missing_edges(self) -> int:
+        """Returns the number of missing edges"""
+        return len(self.missing_edges)
+
+    @property
+    def num_extra_missing_edges(self) -> int:
+        """Returns the number of missing edges that are not incident
+        with a missing node"""
+        return len(self.extra_missing_edges)
+
+    def neighbors(
+        self,
+        v: ZNode,
+        nbr_kind: str | None = None,
+    ) -> set[ZNode]:
+        """Returns the neighbours of v when restricted to nbr_kind"""
+        if not isinstance(v, ZNode):
+            raise TypeError(f"Expected v to be ZNode, got {v}")
+        if v in self.missing_nodes:
+            return {}
```

Comment:
I assume this should return an empty set (`{}` creates an empty dict).
```suggestion
            return set()
```

Reply:
Done in 5ef99a2bd9751c3565b7897ae4a5280abae10a3b